xen-bugtool

```diff
@@ -572,20 +602,64 @@ def include_inventory(archive, dir):
     archive.add_path_with_data(construct_filename(dir, "inventory.xml", {}), info)
 
 
-def collect_data(subdir, archive):
-    process_lists = {}
+def run_procs_and_capture_collected_output(collect, subdir, archive):
+    """Prepare and run processes defined in the passed bugtool data dictionary.
 
-    for (k, v) in data.items():
+    :param collect (dict): Dict with instructions on which processes to run.
+    :param subdir (str): The subdirectory where the output files will be stored.
+    :param archive (obj): The archive object used to store the output files.
+    """
+
+    # Prepare the process_list dictionary
+    process_lists = {}
+    for k, v in collect.items():
         name = construct_filename(subdir, k, v)
         cap = v['cap']
-        filename = v.get('filename')
         if "cmd_args" in v:
             v['output'] = StringIOmtime()
             if cap not in process_lists:
                 process_lists[cap] = []
-            process_lists[cap].append(ProcOutputAndArchive(v['cmd_args'], caps[cap][MAX_TIME], name, archive, v))
-        elif filename and (filename.startswith('/proc/') or
-                           filename.startswith('/sys/')):
+            process_lists[cap].append(
+                ProcOutputAndArchive(
+                    v["cmd_args"],
+                    caps[cap][MAX_TIME],
+                    name,
+                    archive,
+                    v,
+                )
+            )
+
+    # Run the processes in the process_list dictionary in parallel
+    run_procs(process_lists.values())
+
+    # collect all output from processes and free the allocated memory
+    for k, v in data.items():
+        if 'output' in v:
+            archive.add_path_with_data(construct_filename(subdir, k, v), v["output"])
+            v['md5'] = md5sum(v)
+            del v['output']
+
+
+def collect_data(subdir, archive):
+    """Collect all requested data and archive it in the passed output archive
+
+    :param subdir: The toplevel directory in which to store the output files.
+    :param archive: The archive object used to store the output files.
+    """
+    # Run processes first as some (rrd-cli save_rrds) may create/update files:
+    run_procs_and_capture_collected_output(data, subdir, archive)
+
+    # Afterwards, traverse the directory specifications for files to add
+    traverse_directory_specifications(directory_specifications, entries)
+
+    # Then, loop over the files which were found and
+    for k, v in data.items():
```

Comment:
Given this usage the traverse function could be an iterator that calls 'yield' whenever it finds something noteworthy. But its fine this way too.

Reply:
Yes, this is a great idea! This idea would possibly allow interleaving the directory traversal with the data collection, triggered by yielding each file at the moment it is found, which should lead to possibly more effective (faster) data collection, and file collection and would be interlinked, possibly raising chances of getting new files right when they are getting created.